### DISCUSSION TITLE
Correct way to remotely restart a computer using System.Management.Automation

### DISCUSSION THREAD

**Discussion by goffries** [PowerShell/PowerShell#16529] [Q&A]
Hi,

I'm using the System.Management.Automation and System.Management.Automation.Runspaces packages in an attempt to remotely restart a computer, and then wait for that computer to become available again. However whatever i do results in the runspace being in a broken state. Is there any guidance on how to do this in C# using the latest versions of the System.Management.Automation packages?

Sorry if this is too cryptic/vague, I'm fairly new to writing anything in C#/.net and any help would be really appreciated.

**Comment by kasini3000** [comment]
my suggestion is:
Don't use System.Management.Automation.Runspaces
Don't go wait for that computer to become available again
Instead, write two methods:
1 Send computer restart command through psremote
2 Use the powershell command test-wsman and check the windows startup time to determine whether windows restarts normally.

----

You can also write the console program in c# without using the .net class, call powershell.exe, and run the script.
I have written the script for remote execution of commands for you. it's here:
https://gitee.com/chuanjiao10/kasini3000/blob/master/k_run_ip_port_w2w_user_pwd.ps1
You only need to write the logic of the business ps1 script.

**Comment by jborean93** [comment]
You can send the shutdown command with a delay then close down the runspace before it is shut down. From there it's a matter of waiting until the host is back online with some sort of repeated attempts at opening a new runspace. You can probably add further logic to detect once the port you need (5985, 5986) is back online and listening for requests before trying to open the Runspace.
